src/python/pants/backend/native/subsystems/native_build_step_settings_base.py

```diff
@@ -4,37 +4,114 @@
 
 from __future__ import absolute_import, division, print_function, unicode_literals
 
+from pants.backend.native.subsystems.native_build_settings import NativeBuildSettings
 from pants.backend.native.subsystems.utils.mirrored_target_option_mixin import \
   MirroredTargetOptionMixin
 from pants.subsystem.subsystem import Subsystem
+from pants.util.memo import memoized_property
 
 
 class NativeBuildStepSettingsBase(Subsystem, MirroredTargetOptionMixin):
 
+  options_scope = 'native-build-step-settings-base'
+
   mirrored_option_to_kwarg_map = {
     'fatal_warnings': 'fatal_warnings',
+    'compiler_option_sets': 'compiler_option_sets',
   }
 
+  @classmethod
+  def subsystem_dependencies(cls):
+    return super(NativeBuildStepSettingsBase, cls).subsystem_dependencies() + (
+      NativeBuildSettings.scoped(cls),
+    )
+
+  @memoized_property
+  def _native_build_settings(self):
+    return NativeBuildSettings.scoped_instance(self)
+
   @classmethod
   def register_options(cls, register):
     super(NativeBuildStepSettingsBase, cls).register_options(register)
 
-    # TODO: implement compiler_option_sets as an interface to platform/host-specific optimization
-    # flags!
+    # TODO: Deprecate `--fatal-warnings` in a follow-up revision.
     register('--fatal-warnings', type=bool, default=True, fingerprint=True, advanced=True,
              help='The default for the "fatal_warnings" argument for targets of this language.')
+    register('--compiler-option-sets-enabled-args', advanced=True, type=dict, fingerprint=True,
+             default={}, help='Extra compiler args to use for each enabled option set.')
+    register('--compiler-option-sets-disabled-args', advanced=True, type=dict, fingerprint=True,
+             default={}, help='Extra compiler args to use for each disabled option set.')
 
   def get_fatal_warnings_value_for_target(self, target):
     return self.get_target_mirrored_option('fatal_warnings', target)
 
+  def get_merged_compiler_options_for_target(self, target, opt_set_enabled_dicts, opt_set_disabled_dicts):
+    """Merge compiler option sets for a native target into a list of compiler flags."""
+    compiler_option_sets = self._native_build_settings.get_target_mirrored_option(
+        'compiler_option_sets', target)
+    compiler_options = set()
+
+    # Set values for enabled options.
+    if compiler_option_sets:
```

Comment:
Below you directly access the `compiler_option_sets` variable, which implies that it is not `None` within this method... if that's the case, it should be fine to just have this loop at the top level rather than guarded by the `if`.

Reply:
✅